### PR TITLE
Forced load imbalance

### PR DESCRIPTION
**Pull request by nanz-nv**
# What does this PR do ?
This PR tries to add a utility to force load imbalance among experts. dev PR [#2917](https://github.com/NVIDIA/Megatron-LM/pull/2917)

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by Phlip79** [comment]
@nanz-nv please fix test error

File: megatron/core/transformer/transformer_config.py [hunk 1]
```diff
@@ -655,6 +655,13 @@ class TransformerConfig(ModelParallelConfig):
     """[Experimental] Force load balancing with random logits for MoE router, supports naive topk 
     and group-limited topk. This is an experimental feature and only for benchmark."""
 
+    moe_router_force_biased: Optional[float] = None
```
Comment (yashaswikarnati) [inline megatron/core/transformer/transformer_config.py:721]: some nits; 
1) could we make this more explicit in the naming - moe_router_expert_bias_force_random
2) curious why this is experimental

File: megatron/core/transformer/moe/moe_utils.py [hunk 2]
```diff
@@ -1224,6 +1228,54 @@ def apply_random_logits(logits: torch.Tensor) -> torch.Tensor:
     return RandomSTE.apply(logits)
 
 
+@internal_api
+class RandomSTEShared(torch.autograd.Function):
+    """
+    STE that generates random values with shared seed across all ranks.
```
Comment (jaredcasper) [inline megatron/core/transformer/moe/moe_utils.py]: Forgive my ignorance, what is an STE?

Comment (Victarry) [reply megatron/core/transformer/moe/moe_utils.py]: It means [Straight-Through Estimator](https://www.emergentmind.com/topics/straight-through-estimator-ste), which directly returns the output gradients as the input gradients in non-differential cases.
In this function, we want to replace the logits with a random tensor but still make the gradient flow as normal.

File: megatron/core/transformer/transformer_config.py (hunk 1 above)
Comment (nanz-nv) [reply megatron/core/transformer/transformer_config.py:721]: Thanks for the feedback. Because this knob has too modes, and the word random might be misleading, so I decided to keep the name as is for now.
For the second one, I have removed the experimental label. It was to follow the force-load-blance knob

**Comment by ericharper** [comment]
@nanz-nv , why is the argument to add this to the CLI in the dev PR but not in the main PR?

```
    group.add_argument('--moe-router-force-biased', type=float, default=None,
```

File: megatron/core/transformer/moe/moe_utils.py (hunk 2 above)
Comment (jaredcasper) [reply megatron/core/transformer/moe/moe_utils.py]: Sorry, it took a while to get back to this, can you add that to the comment here? I'm sure I'm not the only one that won't immediately know what STE means and what it does.

**Comment by nanz-nv** [comment]
> moe-router-force-biased



> @nanz-nv , why is the argument to add this to the CLI in the dev PR but not in the main PR?
> 
> ```
>     group.add_argument('--moe-router-force-biased', type=float, default=None,
> ```

@ericharper for some reason I cannot reply to your comment. I don't think this is in dev branch either. There is a code change that automatically pass arguments to corresponding transformer_config, so it is not needed anymore.

File: megatron/core/transformer/moe/moe_utils.py (hunk 2 above)
Comment (nanz-nv) [reply megatron/core/transformer/moe/moe_utils.py]: Comment added to match that of existing RandomSTE class.
